File: examples/3-outputs/r/shinyreact.R
library(shiny)

barePage <- function(..., title = NULL, lang = NULL) {
  ui <- list(
    shiny:::jqueryDependency(),
    if (!is.null(title)) tags$head(tags$title(title)),
    ...
  )
  attr(ui, "lang") <- lang
  ui
}

#' Reactively render arbitrary JSON object data.
#'
#' This is a generic renderer that can be used to render any Jsonifiable data.
#' It sends the data to the client-side and let the client-side code handle the
#' rendering.
renderObject <- function(
  expr,
  env = parent.frame(),
  quoted = FALSE,
  outputArgs = list(),
  sep = " "
) {
  func <- installExprFunction(expr, "func", env, quoted, label = "renderObject")

  createRenderFunction(
    func,
    function(value, session, name, ...) {
      value
    },
    function(...) {
      stop("Not implemented")
    },
    outputArgs
  )
}

File: examples/1-hello-world/py/shinyreact.py
from __future__ import annotations

from shiny import ui
from shiny.html_dependencies import shiny_deps
from shiny.types import Jsonifiable
from shiny.render.renderer import Renderer


def page_bare(*args: ui.TagChild, title: str | None = None, lang: str = "en") -> ui.Tag:
    return ui.tags.html(
        ui.tags.head(ui.tags.title(title)),
        ui.tags.body(shiny_deps(False), *args),
        lang=lang,
    )


class render_object(Renderer[Jsonifiable]):
    """
    Reactively render arbitrary JSON object.

    This is a generic renderer that can be used to render any Jsonifiable data.
    It sends the data to the client-side and let the client-side code handle the
    rendering.

    Returns
    -------
    :
        A decorator for a function that returns a Jsonifiable object.

    """

    def __init__(
        self,
        _fn: Optional[ValueFn[str]] = None,
    ) -> None:
        super().__init__(_fn)

    async def transform(self, value: str) -> Jsonifiable:
        return value
